File: Easy/Smallest window containing 0, 1 and 2.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
// User function Template for C++

class Solution
{
  public:
    int smallestSubstring(string S)
    {
        // Code here
        int n = S.length();
        
        queue<int> z;
        queue<int> u;
        queue<int> d;
        
        for (int i=0; i<n; i++)
        {
            if      (S[i] == '0') z.push(i);
            else if (S[i] == '1') u.push(i);
            else if (S[i] == '2') d.push(i);
        }
        
        if (z.size() == 0 || u.size() == 0 || d.size() == 0) return -1;
        
        int ans = INT_MAX;
        int pz = -1;
        int pu = -1;
        int pd = -1;
        
        for (int i=0; i<n; i++)
        {
            if (pz < i)
            {
                if (z.empty()) continue;
                pz = z.front();
                z.pop();
            }
            
            if (pu < i)
            {
                if (u.empty()) continue;
                pu = u.front();
                u.pop();
            }
            
            if (pd < i)
            {
                if (d.empty()) continue;
                pd = d.front();
                d.pop();
            }
            
            int inizio = min(pz, min(pd, pu));
            int fine = max(pz, max(pd, pu));
            
            int finestra = fine-inizio + 1;
            
            ans = min(ans, finestra);
            
        }
        

        if (ans == INT_MAX) return -1;
        return ans;
    }
};


//{ Driver Code Starts.

int main() {
    int t;
    cin >> t;
    while (t--) {
        string S;
        cin >> S;
        Solution ob;
        cout << ob.smallestSubstring(S);
        cout << endl;
    }
}
// } Driver Code Ends

File: Easy/Smallest window containing 0, 1 and 2.py
#User function Template for python3
import sys

class Solution:
    def smallestSubstring(self, S):
        # Code here
        n = len(S)
        ans = sys.maxsize
        
        left = 0
        right = 0
        
        m = {"0": 0, "1": 0, "2": 0}
        
        while right < n:
            m[S[right]] += 1
            
            while m["0"] and m["1"] and m["2"]:
                l = right - left + 1
                ans = min(ans, l)
                m[S[left]] -= 1
                left += 1
            right += 1
        
        if ans < sys.maxsize:
            return ans
        
        return -1


#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__':
	t=int(input())
	for i in range(t):
		S = input()
		ob = Solution()
		ans = ob.smallestSubstring(S)
		
		print(ans)



# } Driver Code Ends
